feat: Support display color space interop IDs in I/O

brechtvl opened the pull request:
## Description

* Display referred interop IDs are now chosen over scene referred ones by
  default on file read. This more closely matches the standards for CICP, ICC
  and other color metadata. Most images will now have `oiio:ColorSpace` set to
  `srgb_rec709_display` rather than `srgb_rec709_scene` on file read.

* To change the behavior to prefer scene default (as may be appropriate for
   textures), set `oiio:ImageStateDefault` to `scene` in `config` for
   `ImageInput.open` or with `oiiotool --iconfig`.

* There is also a new `ImageCache` attribute `image_state_default`. The default
  value here is also `display` as the `ImageCache` is used through OIIO for various
  purposes, but specifically for texture systems `scene` may be a better choice.

* Both `srgb_rec709_scene` and `srgb_rec709_display` are now (distinct) built-in
  color spaces. The built-in `sRGB` color space name remains an alias of
  `srgb_rec709_scene`.

  Documentation was updated to discourage the use of `sRGB` in favor of
  the more specific interop IDs.

* Previously only `srgb_rec709_scene` was recognized as sRGB for file metadata
  and display, now `srgb_rec709_display` and `g22_rec709_display` are treated as
  sRGB as well.

  The reason for `g22_rec709_display` behavior is that this type of display is
  often used to correct for the discrepancy where images are encoded as sRGB but
  usually decoded as gamma 2.2 by the physical display. By encoding it as gamma
  2.2 and claiming it's sRGB the transfer functions cancel out exactly.

* `g24_rec709_display` is now recognized as having gamma 2.4, and the name
  `g24_rec709_scene` was replaced with `ocio:g24_rec709_scene` since the
  former is not an official interop ID.

* This change also fixes an issue where tga and rla were incorrectly using
  `g22_rec709` and `g18_rec709` without `_scene` suffix, and not handling
  `g24_rec709`.

* `cio:itu709_rec709_scene` and `ocio:lin_ciexyzd65_display` were added to complete
  the list of interop IDs in OCIO configs that match a CICP.

* `ColorConfig` now includes inactive colorspaces. In older ACES configs (like
  v2.1.0 bundled with OpenColorIO 2.3), the display color spaces are inactive.
  But we now heavily rely on them.

## Tests

* Many test outputs were changed due to the new display referred default.
* Added test for `g22_rec709_display` behavior.
* Added test for `color:prefer_image_state`.

## Checklist:

- [x] I have read the [contribution guidelines](https://github.com/AcademySoftwareFoundation/OpenImageIO/blob/main/CONTRIBUTING.md).
- [x] I have updated the documentation, if applicable. (Check if there is no
  need to update the documentation, for example if this is a bug fix that
  doesn't change the API.)
- [x] I have ensured that the change is tested somewhere in the testsuite
  (adding new test cases if necessary).
- [x] If I added or modified a C++ API call, I have also amended the
  corresponding Python bindings (and if altering ImageBufAlgo functions, also
  exposed the new functionality as oiiotool options).
- [x] My code follows the prevailing code style of this project. If I haven't
  already run clang-format before submitting, I definitely will look at the CI
  test that runs clang-format and fix anything that it highlights as being
  nonconforming.

brechtvl committed:
feat: Support display color space interop IDs in I/O

* Display referred interop IDs are now chosen over scene referred ones by
  default on file read. This more closely matches the standards for CICP, ICC
  and other color metadata. Most images will now have oiio:ColorSpace set to
  srgb_rec709_display rather than srgb_rec709_scene on file read.

  There is a new global attribute color:prefer_image_state that when changed
  from the default "display" to "scene" will prefer scene referred color spaces
  as before.

* Both srgb_rec709_scene and srgb_rec709_display are now (distinct) built-in
  color spaces. The built-in "sRGB" color space name was changed to be an alias
  of srgb_rec709_display rather than srgb_rec709_scene.

* Previously only srgb_rec709_scene was recognized as sRGB for file metadata
  and display, now srgb_rec709_display and g22_rec709_display are treated as
  sRGB as well.

  The reason for g22_rec709_display behavior is that this type of display is
  often used to correct for the discrepancy where images are encoded as sRGB but
  usually decoded as gamma 2.2 by the physical display. By encoding it as gamma
  2.2 and claiming it's sRGB the transfer functions cancel out exactly.

* g24_rec709_display is now recognized as having gamma 2.4, and the name
  g24_rec709_scene was replaced with ocio:g24_rec709_scene since the
  former is not an official interop ID.

* cio:itu709_rec709_scene and ocio:lin_ciexyzd65_display were added to complete
  the list of interop IDs in OCIO configs that match a CICP.

* ColorConfig now includes inactive colorspaces. In older ACES configs (like
  v2.1.0 bundled with OpenColorIO 2.3), the display color spaces are inactive.
  But we now heavily rely on them.

Signed-off-by: Brecht Van Lommel <brecht@blender.org>

brechtvl committed:
feat: Refactor writing color space metadata to unify behavior

This adds new functions to get and set colorspaces in a ImageSpace, to
make logic between file formats more consistent.
* pvt::set_colorspace_srgb and pvt::is_colorspace_srgb
* pvt::get_colorspace_rec709_gamma
* pvt::get_colorspace_icc_profile
* pvt::set_colorspace_cicp and pvt::get_colorspace_cicp

This fixes an issue where tga and rla were incorrectly using g22_rec709
and g18_rec709 without _scene suffix, and not handling g24_rec709.

There is existing consistency in that some file formats assume an empty
oiio:ColorSpace to mean sRGB, and some don't. This inconsistency is
preserved.

Signed-off-by: Brecht Van Lommel <brecht@blender.org>

lgritz commented on src/libOpenImageIO/color_ocio.cpp:
```diff
@@ -2864,11 +2870,14 @@ pvt::get_colorspace_rec709_gamma(const ImageSpec& spec)
-    // Note g24_rec709_scene is not a standard interop ID
-    else if (colorconfig.equivalent(colorspace, "g24_rec709_scene"))
+    // Note g24_rec709_scene is not a standard interop
+    else if (colorconfig.equivalent(colorspace, "g24_rec709_scene")
```
Is g24_rec709_scene not a real interop tag?

lgritz replied:
So here I am torn about this prefixing. Does OCIO recognize "oiio:g24_rec709_scene", or does this mean that something elsewhere in OIIO we need to look for, and strip out, the prefixes any time it needs to actually call OCIO? That seems pretty ugly, too. 

Convince me otherwise, but I'm inclined to think that the "oiio:ColorSpace" attrib can have any unadorned OCIO color space name, as long as it's either in the current config or the built-in studio config. Or am I thinking of it wrong?

doug-walker replied:
The interop ID description [is here](https://docs.google.com/document/d/1T94lYbis9uCskL_ZEMxGBF2JryLfZnjxlEoNgRHZzBE/edit?usp=sharing). The "g24_rec709_scene" name is not on the [CIF texture ID list](https://github.com/AcademySoftwareFoundation/ColorInterop/blob/main/Recommendations/01_TextureAssetColorSpaces/TextureAssetColorSpaces.md#summary-table--overview-of-the-recommendations). Therefore it needs a namespace prefix. The prefix we use in the OCIO configs for ACES is "ocio". You would never strip that out, it's part of the interop ID. Zach is trying to keep OIIO in sync with the usage in the released OCIO configs.

doug-walker replied:
> Else, strip the namespace prefix and the ":", and see if the remaining string matches a name or alias of a color space in the current config.

That was indeed at one point in the interop ID document I link to above. However, it was decided to remove it in the interest of making things simpler. As the document stands now, there is no search without the namespace part of the ID.

lgritz replied:
Is the built-in studio config a strict superset of the default OCIO cg config these days? Should we make the studio config be the one that OIIO uses as a default / backstop? I'm not sure if there are performance implications to that. 

But the reason why the studio config is not the default is only because a couple years ago when I was looking at it, the main thing the studio config added was cameras that I thought "naive" users (those not working in a production environment with its own config) would not need. But if it has evolved to the point where a bunch of color spaces that people might encounter in files found in the wild, but that are are not pro-camera-specific (i.e. stuff that any old CICP tags might implicate) are in the studio config but not on the default cg config, maybe we should re-evaluate the choice?

zachlewis replied:
Welp, I'd like to imbue the ColorConvert IBA with the ability to "do the right thing" given a known "builtin interop_id" (i.e., the set of interop_id values associated  with each Studio config color space), regardless of whether the current ColorConfig actually contains a color space matching that name or associated definition. 

In other words, the entire set of Studio config color spaces will always be accessible to ColorConvert (except in rare cases where OCIO's heuristics cannot ascertain a way to bridge the active ColorConfig and the builtin Studio config). This means there's a reserved set of ColorConfig-agnostic strings that will reliably work with ColorConvert:

```python
import PyOpenColorIO as ocio

def get_builtin_interop_ids() -> list[str]:
  return [i.getInteropID() for i in ocio.Config.CreateFromFile("ocio://studio-config-latest").getColorSpaces()]
```
```python
print(sorted(get_builtin_interop_ids(), key = lambda s: s.startswith("ocio:")))
```
```
['srgb_rec709_display', 'g22_rec709_display', 'srgb_p3d65_display', 'srgbe_p3d65_display', 'g26_p3d65_display', 'g24_rec709_display', 'hlg_rec2020_display', 'pq_rec2020_display', 'pq_p3d65_display', 'lin_ap0_scene', 'lin_ap1_scene', 'srgb_rec709_scene', 'g18_rec709_scene', 'g22_rec709_scene', 'srgb_p3d65_scene', 'g22_adobergb_scene', 'srgb_ap1_scene', 'g22_ap1_scene', 'lin_adobergb_scene', 'lin_p3d65_scene', 'lin_rec2020_scene', 'lin_rec709_scene', 'data', 'ocio:acescc_ap1_scene', 'ocio:acescct_ap1_scene', 'ocio:adx10_apd_scene', 'ocio:adx16_apd_scene', 'ocio:applelog_rec2020_scene', 'ocio:arrilogc3_awg3_scene', 'ocio:lin_awg3_scene', 'ocio:arrilogc4_awg4_scene', 'ocio:lin_awg4_scene', 'ocio:bmdfilm5_wg5_scene', 'ocio:lin_bmdwg5_scene', 'ocio:davinci_dwg_scene', 'ocio:lin_dwg_scene', 'ocio:canonlog2_cgamutd55_scene', 'ocio:lin_cgamutd55_scene', 'ocio:canonlog3_cgamutd55_scene', 'ocio:djilog_dgamut_scene', 'ocio:lin_dgamut_scene', 'ocio:vlog_vgamut_scene', 'ocio:lin_vgamut_scene', 'ocio:redlog3g10_rwg_scene', 'ocio:lin_rwg_scene', 'ocio:slog3_sgamut3_scene', 'ocio:slog3_sgamut3cine_scene', 'ocio:slog3_sgamut3venice_scene', 'ocio:slog3_sgamut3cinevenice_scene', 'ocio:lin_sgamut3_scene', 'ocio:lin_sgamut3cine_scene', 'ocio:lin_sgamut3venice_scene', 'ocio:lin_sgamut3cinevenice_scene', 'ocio:g24_rec709_scene', 'ocio:itu709_rec709_scene']
```

lgritz replied:
Indeed, I'm in favor of a general heuristic of "if it's not in the active config, try the broadest built-in config we have, and it's only an error if it still can't figure it out at that point."

doug-walker commented on src/libOpenImageIO/color_ocio.cpp:
```diff
@@ -2139,15 +2196,23 @@ ColorConfig::get_color_interop_id(string_view colorspace) const
         }
     }
-    return "";
+    return other_interop_id;
```
Was other_interop_id properly initialized?

doug-walker commented on src/libOpenImageIO/color_ocio.cpp:
```diff
@@ -2057,6 +2100,43 @@ struct ColorInteropID {
+
     // Scene referred interop IDs first so they are the default in automatic
     // conversion from CICP to interop ID. Some are not display color spaces
     // at all, but can be represented by CICP anyway.
```
Scene IDs are no longer first.

doug-walker commented on src/libOpenImageIO/color_ocio.cpp:
```diff
@@ -581,14 +597,18 @@ ColorConfig::Impl::classify_by_name(CSInfo& cs)
+    if (Strutil::iequals(cs.name, "srgb_rec709_display")
+        || Strutil::iequals(cs.name, "srgb_display")
+        || Strutil::iequals(cs.name, "sRGB - Display")
         || Strutil::iequals(cs.name, "sRGB")) {
```
Is moving "sRGB" to srgb_rec709_display going to be disruptive? Just wondering if it should stay as srgb_rec709_scene, or be based on prefer_image_state, as it does not seem directly related to CICP.

lgritz replied:
I guess the question is: when a naive user asks to convert something to or from "sRGB", do they mean srgb_rec709_display or srgb_rec709_scene?

(I still struggle to understand under which practical, in-the-wild situations those would yield different pixel values, except for expert users who have defined an OCIO config that has an srgb color space with tone mapping, right?)

brechtvl replied:
I think in general using `sRGB` should not be recommended anymore, and I could update the documentation for that. With that in mind, keeping it as `srgb_rec709_scene` would give better backwards compatibility for something that we are phasing out. The reason I set it to `srgb_rec709_display` was to match the default interpretation of sRGB image files. I'm happy to change it if backwards compatibility is preferred for this case.

---

My understanding is that with the ACES and other well designed configs, there is no difference for direct conversions between color spaces for `srgb_rec709_display` and `srgb_rec709_scene`.

The difference between these spaces is the default view transform for commands like `--ociodisplay`. For scene referred it will apply e.g. the ACES view transform, while for display referred it will effectively skip it and do a direct conversion to the chosen display color space.

These assumptions break with configs that:
* Incorrectly set an interop ID as an alias on a color space that does not match the CIF recommendation.
* Set `default_view_transform` (the connection between scene and display spaces, not the default view transform mentioned above) to something that's arguably invalid.

brechtvl replied:
The most backwards compatible is `sRGB` = `srgb_rec709_scene`, so I'll change it back to that. (Even if in your example it would be wrong, but wrong the same way as before, and only for certain OCIO configs that are probably uncommon.)

I'll write some example commands in the comments for posteriority.

doug-walker replied:
Here is more detail about the default_view_transform. OCIO has two primary methods of converting between color spaces: ColorSpaceTransform (which is for converting between color spaces of the same image state) and DisplayViewTransform (which is for converting between color spaces of different image states). However, those transforms try to be forgiving and not throw an exception if someone tries to use, e.g. a ColorSpaceTransform to convert from a scene-referred to a display-referred color space. In those scenarios, a View Transform must be inserted to allow the processing. That's the purpose of the default_view_transform. (It is not the default view, for which it is sometimes mistaken due to the similarity in the naming.)

It could plausibly be any View Transform in the config. For example, someone could make it the same as the default view (i.e., the first entry in the active_views) and that would be a perfectly fine choice. That would likely even be a better choice in some industries, but Un-tone-mapped makes the most sense for CG since it makes srgb_rec709_scene and srgb_rec709_display behave the same in most, though not all, circumstances.

I agree that this is something that is best decided by the config author and should not be modified by applications or end-users.

But I will stress that if the color space tag is correctly assigned, and the application is using the appropriate type of transform for the task at hand (ColorSpace vs. DisplayView), then the default_view_transform is never used. It is a fallback.

doug-walker reviewed (commented):
I did a quick pass through this one. I'm still an OIIO novice, but it seems like there are a lot of nice improvements in here.

brechtvl committed:
Address comments, update docs to discourage using "sRGB" colorspace

Signed-off-by: Brecht Van Lommel <brecht@blender.org>

brechtvl commented:
Regarding the current status for this PR. I'm not currently planning further changes. There's more refactoring that could done, for deferred color space determination and more unified code. But that's probably best left for another PR?

I think the main decisions to be taken are:
* [x] Defaut to scene or display referred? (Consensus seems to be display.)
* [ ] Write CICP and ICC even for scene referred color spaces? (In my opinion, yes)
* [ ] Is the special `g22_rec709_display` handling a good thing? Or too confusing or undesirable for users with a Gamma 2.2 workflow? Or we won't have a good answer soon, so remove the exception and defer the decision? (No opinion)
* [ ] Should "sRGB" be `srgb_rec709_scene` (backwards compatibility) or `srgb_rec709_display` (consistency with default image load) or even be controlled by `color:prefer_image_state`? (No opinion)

brechtvl committed:
Set "sRGB" as alias for "srgb_rec709_scene" again

Signed-off-by: Brecht Van Lommel <brecht@blender.org>

brechtvl committed:
Merge branch 'main' into display-interop-id-io

Signed-off-by: Brecht Van Lommel <brecht@blender.org>

brechtvl committed:
Make the default image state an image input config and image cache attribute

Rather than a global attribute, so that different parts of a multithreaded
application can use different behavior.

Signed-off-by: Brecht Van Lommel <brecht@blender.org>

brechtvl committed:
Merge branch 'main' into display-interop-id-io

Signed-off-by: Brecht Van Lommel <brecht@blender.org>

brechtvl commented:
I have made changes so preferring scene or display referred image state can be controlled per `ImageInput` and `ImageCache`, as I think that's the control you'd need in practice in multithreaded applications that may handle different types of images.

This required changes in many files though, so it's. a bit ugly right but I could not see a nice way to do it without breaking the API significantly. If deferred color space resolution is added at some point this can probably be simplified.

zachlewis commented on src/doc/imagebufalgo.rst:
```diff
@@ -2950,14 +2952,15 @@ Color space conversion
           ImageBuf Src ("tahoe.exr");
-          ImageBuf Dst = ImageBufAlgo::ociodisplay (Src, "sRGB", "Film", "lnf",
-                                                    "", true, "SHOT", "pe0012");
+          ImageBuf Dst = ImageBufAlgo::ociodisplay (Src, "srgb_rec709_scene",
+                                                    "Film", "lnf", "", true,
+                                                    "SHOT", "pe0012");
```
In this example, "sRGB" is actually the name of the display, not the source color space, so it doesn't make sense to replace with "srgb_rec709_scene" here... however, I wouldn't mind replacing "lnf"  (the source color space) with "srgb_rec709_scene". 
```suggestion
          ImageBuf Dst = ImageBufAlgo::ociodisplay (Src, "sRGB",
                                                    "Film", "srgb_rec709_scene", "", true,
                                                    "SHOT", "pe0012");
```

zachlewis commented on src/doc/imagebufalgo.rst:
```diff
@@ -2950,14 +2952,15 @@ Color space conversion
           Src = ImageBuf("tahoe.jpg")
-          Dst = ImageBufAlgo.ociodisplay (Src, "sRGB", "Film", "lnf",
-                                          "", True, "SHOT", "pe0012")
+          Dst = ImageBufAlgo.ociodisplay (Src, "srgb_rec709_scene", "Film",
+                                          "lnf", "", True, "SHOT", "pe0012")
```
Same as above, "sRGB" is supposed to be the name of the display.

This suggestion is a twist on the above; it resolves the default view for the default display when the input color space ~= "srgb_rec709_scene"
```suggestion
          Dst = ImageBufAlgo.ociodisplay (Src, "", "",
                                          "srgb_rec709_scene", "", True, "SHOT", "pe0012")
```

zachlewis commented on src/doc/pythonbindings.rst:
```diff
@@ -3660,7 +3661,7 @@ Color manipulation
 
         Src = ImageBuf ("tahoe.exr")
-        Dst = ImageBufAlgo.ociodisplay (Src, "sRGB", "Film", "lnf",
+        Dst = ImageBufAlgo.ociodisplay (Src, "srgb_rec709_scene", "Film", "lnf",
```
Another twist: resolve the default view for the "sRGB" display when the input color space ~= "srgb_rec709_scene"
```suggestion
        Dst = ImageBufAlgo.ociodisplay (Src, "sRGB", "", "srgb_rec709_scene",
```

zachlewis commented on src/dpx.imageio/dpxoutput.cpp:
```diff
@@ -436,18 +438,14 @@ DPXOutput::prep_subimage(int s, bool allocate)
+    else if (gamma != 0.0f)
         m_transfer = dpx::kUserDefined;
-    else if (colorconfig.equivalent(colorspace, "KodakLog"))
+    else if (spec_s.get_string_attribute("oiio:ColorSpace") == "KodakLog")
```
This doesn't seem very reliable. I'm fine with leaving the behavior intact for this PR -- I'd also be fine with removing this clause entirely for now. 

In the future, we could use the `oiio:ColorSpace` color space's `encoding` attribute to map to various dpx:Transfer constants. (DPX metadata isn't very reliable in my experience, beyond distinguishing "log" from "video" encodings)

zachlewis commented on src/doc/imagebufalgo.rst:
```diff
@@ -2849,12 +2849,14 @@ Color space conversion
 
           Src = ImageBuf("tahoe.jpg")
-          Dst = ImageBufAlgo.colorconvert (Src, "sRGB", "acescg", True)
+          Dst = ImageBufAlgo.colorconvert (Src, "srgb_rec709_scene", "acescg",
```
```suggestion
          Dst = ImageBufAlgo.colorconvert (Src, "srgb_rec709_scene", "lin_ap1_scene",
```
(config-agnostic behavior)

zachlewis commented on src/doc/stdmetadata.rst:
```diff
@@ -149,9 +149,12 @@ Color information
-    - `"g22_rec709_scene"` : Rec709/sRGB primaries, but using a response curve
+    - `"srgb_rec709_display"` : Using standard (piecewise) sRGB response and
+      primaries. The token `"sRGB"` is treated as a synonym, but it is
+      recommended to use the more specific interop ID.
```
I'd like to deprecate treating "sRGB" as a synonym for either "srgb_rec709_display" or "srgb_rec709_scene"

zachlewis reviewed (commented):
Thanks for your patience, Brecht! 
I apologize, I started this review three weeks ago; I didn't realize I hadn't submitted my approval! The only stuff I have comments on pertain to documentation. 

I think I understand what you're trying to do with `oiio:imageStateDefault` etc -- if I'm understanding correctly, a use case might be a file-browser UI for picking scene-referred diffuse texture maps; and the goal is to prevent downstream handlers from automatically applying an inverse display-view transform instead of a simple color space transform when applying a deferred conversion to the rendering space. That's a totally reasonable use case, and I think exposing an attribute for controlling this on the spec-level is a pretty good way to handle this. 

Maybe, in the future, we can use the same attribute to convey information about a specific OCIO View to use by default for inverse display-view transforms.

Anyway, thanks for your work on this, and, again, I apologize for my super-slow response.
